### PR TITLE
Ecal shape db

### PR DESCRIPTION
**Pull request by theofil**
This is about an improved ECAL Shape Class to be used upon digitization. It supports sub nanosecond precision in the ECAL signal pulses and is fully configurable from the DB, in view of the Phase II upgrade where the detector will have a shorter pulse shape. More details can be found in my ECAL DPG presentation:

https://indico.cern.ch/event/731948/

The new shape class supports also the old style Phase-I logic that makes no attempt to fetch information from the DB. In this pull request the code is committed with the DB usage currently de-activated, to allow easier approval of the PR. A second PR is foreseen once this PR is approved (and thus also the new DB Ecal Object can go in the global tag), to actually enable the DB configuration and read the conditions from the corresponding global tag.

I have validated the resulting output with a small scale ZEE MC production as well as a dedicated tool that instantiates ad-hoc pulses.

### CODE REVIEW

**Comment by slava77** [comment]
there are 4K lines of text files with generated numbers.
Why are these numbers not in a DB already?

**Comment by theofil** [comment]
@slava77 if you are referring to those files:

CondTools/Ecal/python/APD_SimPulseShape.txt  
CondTools/Ecal/python/EB_SimPulseShape_PhaseII.txt  
CondTools/Ecal/python/EB_SimPulseShape.txt  
CondTools/Ecal/python/EE_SimPulseShape.txt  

those are use to create the sqlite3 file, which is later-on passable to create the DB conditions. For me, they are not essential to stay in the release and can be removed, if needed. But for future reference and to allow other developers to see how the code works, it might be good to have them. However, to code works without them.

**Comment by theofil** [comment]
I see that I need to run the

```
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208/git-diff.patch
e.g. curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208/git-diff.patch | patch -p1

```
and update the PR. Should I do that already now, or wait for all the release managers to comment in present version of the PR and submit the second PR hopefully integrating all comments received ?

**Comment by slava77** [comment]
On 6/15/18 11:01 AM, Konstantinos Theofilatos wrote:
> I see that I need to run the
> 
> |-code-checks Logs:
> https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208 Code check
> has found code style and quality issues which could be resolved by
> applying a patch in
> https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208/git-diff.patch
> e.g. curl
> https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23600/5208/git-diff.patch
> | patch -p1 |
> 
> and update the PR. Should I do that already now, or wait for all the
> release managers to comment in present version of the PR and submit the
> second PR hopefully integrating all comments received ?
> 

your PR is actually "invisible" to reviewers until code checks pass
(unless someone watches the full repository)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> Reco comparison results: 374 differences found in the comparisons

A breakdown by workflow is from https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-06-17-2300+23600/27130/validateJR/logRootQA.log
```
JR results differ 11 all_mini_OldVSNew_TTbar14TeV2023D17wf20034p0
JR results differ 10 all_OldVSNew_TTbar13TeV2017wf10024p0
JR results differ 7 all_mini_OldVSNew_TTbar14TeV2023D21wf21234p0
JR results differ 24 all_mini_OldVSNew_TTbar13TeVPU2017wf10224p0
JR results differ 18 all_mini_OldVSNew_Higgs200ChargedTauswf9p0
JR results differ 2 all_OldVSNew_TTbar14TeV2023D17wf20034p0
JR results differ 19 all_OldVSNew_TTbar13TeV2018wf10824p0
JR results differ 15 all_mini_OldVSNew_TTbarPUwf25202p0
JR results differ 4 all_OldVSNew_HydjetQB12in2018wf150p0
JR results differ 33 all_OldVSNew_TTbarPUwf25202p0
JR results differ 4 all_OldVSNew_TTbar14TeV2023D19wf20434p0
JR results differ 19 all_OldVSNew_TTbar13TeV2019wf11624p0
JR results differ 20 all_mini_OldVSNew_TTbar13TeVPUppmx2018wf250202p181
JR results differ 42 all_OldVSNew_TTbar13TeVPUppmx2018wf250202p181
JR results differ 26 all_mini_OldVSNew_TTbar13TeV2018wf10824p0
JR results differ 12 all_mini_OldVSNew_TTbar13TeV2019wf11624p0
JR results differ 12 all_mini_OldVSNew_TTbarwf25p0
JR results differ 3 all_OldVSNew_TTbar14TeV2023D21wf21234p0
JR results differ 16 all_mini_OldVSNew_TTbar13TeV2017wf10024p0
JR results differ 10 all_mini_OldVSNew_TTbar14TeV2023D19wf20434p0
JR results differ 14 all_OldVSNew_TTbarwf25p0
JR results differ 12 all_OldVSNew_Higgs200ChargedTauswf9p0
JR results differ 6 all_OldVSNew_ZMM13TeVwf1330p0
JR results differ 33 all_OldVSNew_TTbar13TeVPU2017wf10224p0
JR results differ 2 all_OldVSNew_SingleElectronE120EHCALwf101p0
SUMMARY Reco comparison results: 374 differences found in the comparisons
```

The changes are apparently small and affect only MC. I thought that no changes were expected for existing workflows.

**Comment by theofil** [comment]
@slava77 
I have validated that the pulses will not change to the level of 10^-5, so there would be some numerical differences due to rounding and floating point arithmetic, the EcalShapeClass has updated logic to not assume just 1.0ns sampling for the internally stored shape information. The sampling time is actually dividing or multiplying intermediate products of the shape interpolation logic. 

Having said so, how can I check if the 374 changes are indeed *small* and verify my expectation ?

Also, I see a unit test is failing  

**test EcalIntercalibConstants_O2O_test had ERRORS**

which comes to me as surprise, because I did nothing related to EcalIntercalibConstants_O2O_test.

**Comment by fabiocos** [comment]
@theofil the unit test error is unrelated for what I ca see, and should be fixed by merging #23610

**Comment by slava77** [comment]
On 6/19/18 1:36 AM, Konstantinos Theofilatos wrote:
> @slava77 <https://github.com/slava77>
> I have validated that the pulses will not change to the level of 10^-5,
> so there would be some numerical differences due to rounding and
> floating point arithmetic, the EcalShapeClass has updated logic to not
> assume just 1.0ns sampling for the internally stored shape information.
> The sampling time is actually dividing or multiplying intermediate
> products of the shape interpolation logic.
> 
> Having said so, how can I check if the 374 changes are indeed /small/
> and verify my expectation ?
> 

I usually download the files and look.
The diffs are all in MC, which is suspicious.
The MC tests also have almost an order of magnitude fewer events than
the data tests.
Are your changes affecting digitization in some way?

**Comment by slava77** [comment]
On 6/19/18 5:23 AM, Konstantinos Theofilatos wrote:
> @slava77 <https://github.com/slava77>
> Yes this PR, is related to the Digitization, the pulse shape is what is
> used to generate the Digis. Apart from numerical rounding differences
> the new pulse shape, should be identical to the old, at least this is
> the intention.
> 
> In the meanwhile, I have generated my self the tests by running
> 
> *runTheMatrix.py -l limited -i all*
> 
> but I am not sure where to find in the produced folder the differences
> between the IB before my PR and e.g., the result distribution if my PR
> is applied. Are there DQM style histograms that should be checked ?
> 
> —

Outputs of each workflow with DQM running should have a DQM*.root file,
which can be compared with the baseline.
In this context of tiny differences, you are probably better off to make
the baseline reference as well.

**Comment by theofil** [comment]
Hi,

I've updated the pull request to comply with the CMSSW code style, removing the `using namespace std;`from header files.

By browsing the DQM release monitoring plots, I didn't spot any major issue, apart from the expected small differences in the Ecal Digis/RecHits which I was expecting due to numerical precision/rounding.

see e.g.,

- [https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-06-17-2300+23600/27130/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018/EcalDigisV_EcalDigiTask.html](Ecal Digi Task)
- [https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-06-17-2300+23600/27130/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018/EcalRecHitsV__EcalRecHitsTask_EcalRecHitsTask%20Barrel%20RecSimHit%20Ratio.png](url)

To the best of my knowledge, this is the only major difference that is introduced by this PR.

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc [hunk 1]
```diff
@@ -5,7 +5,6 @@
 #include <algorithm>
 
 #include <iostream>
-using namespace std;
```
Comment (slava77) [inline SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: this is OK in .cc files

Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: even if OK, since it was not needed, I decided to remove it also from the .cc file

**Comment by civanch** [comment]
please test

File: CondFormats/EcalObjects/src/classes.h [hunk 2]
```diff
@@ -222,5 +223,14 @@ namespace CondFormats_EcalObjects {
...
     EcalCondObjectContainer<EcalPulseSymmCovariance> ecalSPCmap; //typedef EcalPulseSymmCovariance
 
+    EcalSimPulseShape ecal_sim_pulse_shapes;
+    std::vector<double> v_barrel_shape;
+    std::vector<double> v_endcap_shape;
+    std::vector<double> v_apd_shape; 
+    double barrel_thresh;
+    double endcap_thresh;
+    double apd_thresh;
+    float time_interval;
```
Comment (slava77) [inline CondFormats/EcalObjects/src/classes.h]: these basic types  don't seem to belong in this file.
I'd expect only EcalSimPulseShape with whatever containers using/needing it.

File: CondTools/Ecal/python/copySimPulseShapeFromFile_cfg.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("TEST")
```
Comment (slava77) [inline CondTools/Ecal/python/copySimPulseShapeFromFile_cfg.py]: i'm not sure about db practices.
In other areas/subsystems, files like this go to the /test directory

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h [hunk 4]
```diff
@@ -57,8 +57,9 @@ class EcalUncalibRecHitWorkerWeights : public EcalUncalibRecHitWorkerRunOneDigiB
                 EcalUncalibRecHitRecWeightsAlgo<EBDataFrame> uncalibMaker_barrel_;
                 EcalUncalibRecHitRecWeightsAlgo<EEDataFrame> uncalibMaker_endcap_;
 
- 		const EEShape testbeamEEShape; // used in the chi2
-                const EBShape testbeamEBShape; // can be replaced by simple shape arrays of floats in the future (kostas)
+                bool useDBShape;
+ 		EEShape testbeamEEShape; // used in the chi2
+                EBShape testbeamEBShape; // can be replaced by simple shape arrays of floats in the future (kostas)
```
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h]: please use a consistent indentation

File: RecoTBCalo/EcalTBRecProducers/interface/EcalTBWeightUncalibRecHitProducer.h [hunk 5]
```diff
@@ -37,8 +37,8 @@ class EcalTBWeightUncalibRecHitProducer : public edm::EDProducer {
     EcalUncalibRecHitRecWeightsAlgo<EBDataFrame> EBalgo_;
     EcalUncalibRecHitRecWeightsAlgo<EEDataFrame> EEalgo_;
 
-    const EEShape testbeamEEShape;  
-    const EBShape testbeamEBShape; 
+    EEShape testbeamEEShape;  
+    EBShape testbeamEBShape;
```
Comment (slava77) [inline RecoTBCalo/EcalTBRecProducers/interface/EcalTBWeightUncalibRecHitProducer.h]: it looks like this change is not needed here because these shapes are never modified after the construction

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h (hunk 4 above)
Comment (theofil) [reply RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h]: the identation should be consistent, in the [raw file](https://raw.githubusercontent.com/theofil/cmssw/ecalrebase/RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h)  it appears consistent

File: CondTools/Ecal/python/copySimPulseShapeFromFile_cfg.py (hunk 3 above)
Comment (theofil) [reply CondTools/Ecal/python/copySimPulseShapeFromFile_cfg.py]: I don't have a strong opinion, this file was initially created by Francesca Cavalari, who took care the creation of the DB format. I only did minor adjustments on the format itself, but not on the files and their structure throughout the various DB related folders

File: RecoTBCalo/EcalTBRecProducers/interface/EcalTBWeightUncalibRecHitProducer.h (hunk 5 above)
Comment (theofil) [reply RecoTBCalo/EcalTBRecProducers/interface/EcalTBWeightUncalibRecHitProducer.h]: Yes. I updated this. Now it has the const. It will be fixed in the next updated of the PR.

File: CondFormats/EcalObjects/src/classes.h (hunk 2 above)
Comment (theofil) [reply CondFormats/EcalObjects/src/classes.h]: thanks for the catch. this is now cleaned up. I will update the PR soon.

**Comment by theofil** [comment]
I updated the PR, incorporating the comments of @slava77 and adding a new unit test to test the Serialization of the EcalSimPulseShape

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23600	4f85358
- changes are in line with the PR description. The reco code is affected only minimally, to allow pick up of the ECAL pulse shapes from DB; the default is not changed.
- jenkins tests pass and comparisons with the baseline show only small differences in the MC plots, which are driven by some numerical precision level changes in the code that affects digitization. There are no changes in the data comparisons.

**Comment by civanch** [comment]
@theofil , @slava77 , differences in jetMET and PF plots are fine for WFs 5.1, 25.0, 10024.0, 12240.0? WHy only in these WFs but not in others for run-1 and run-2? From the description I would expect disagreements with the baseline in newer WFs only.

**Comment by slava77** [comment]
On 6/25/18 3:48 PM, Vladimir Ivantchenko wrote:
> @theofil <https://github.com/theofil> , @slava77
> <https://github.com/slava77> , differences in jetMET and PF plots are
> fine for WFs 5.1, 25.0, 10024.0, 12240.0? WHy only in these WFs but not
> in others for run-1 and run-2? From the description I would expect
> disagreements with the baseline in newer WFs only.
> 

reco comparisons show differences in ~all MC workflows (25.0 and 9.0 in
run1;
10224 and 25202 in run2 etc

12 	 all_OldVSNew_Higgs200ChargedTauswf9p0
4 	 all_OldVSNew_HydjetQB12in2018wf150p0
2 	 all_OldVSNew_SingleElectronE120EHCALwf101p0
7 	 all_OldVSNew_TTbar13TeV2017wf10024p0
8 	 all_OldVSNew_TTbar13TeV2018wf10824p0
14 	 all_OldVSNew_TTbar13TeV2019wf11624p0
33 	 all_OldVSNew_TTbar13TeVPU2017wf10224p0
40 	 all_OldVSNew_TTbar13TeVPUppmx2018wf250202p181
2 	 all_OldVSNew_TTbar14TeV2023D17wf20034p0
4 	 all_OldVSNew_TTbar14TeV2023D19wf20434p0
3 	 all_OldVSNew_TTbar14TeV2023D21wf21234p0
30 	 all_OldVSNew_TTbarPUwf25202p0
14 	 all_OldVSNew_TTbarwf25p0
8 	 all_OldVSNew_ZMM13TeVwf1330p0

**Comment by fabiocos** [comment]
@ggovi @arunhep @lpernie @civanch @rekovic could you please have a look?

**Comment by ggovi** [comment]
+1

**Comment by lpernie** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
attn @kpedro88 

I'd like to have this in the PhaseII dedicated release, which I presume will be 10_2_X.

**Comment by kpedro88** [comment]
assign upgrade

**Comment by kpedro88** [comment]
@rekovic it is more likely that official production will use 10_3_X or later, since the MTD TDR is delayed. However, there may be relvals from 10_2_0.

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc [hunk 6]
```diff
@@ -3271,3 +3292,77 @@ EcalTrivialConditionRetriever::produceEcalSamplesCorrelation( const EcalSamplesC
        back_inserter(ipar->EEG1SamplesCorrelation));
   return ipar;
 }
+
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: remove unnecessary white space

File: CalibCalorimetry/EcalTrivialCondModules/interface/EcalTrivialConditionRetriever.h [hunk 7]
```diff
@@ -219,6 +224,12 @@ class EcalTrivialConditionRetriever : public edm::ESProducer,
   std::vector<double> energyUncertaintyParameters_;
   std::vector<double> energyCorrectionObjectSpecificParameters_;
 
+  double sim_pulse_shape_EB_thresh;
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/interface/EcalTrivialConditionRetriever.h]: use a trailing `_` to denote member variables

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc [hunk 8]
```diff
@@ -3271,3 +3292,77 @@ EcalTrivialConditionRetriever::produceEcalSamplesCorrelation( const EcalSamplesC
...
+( const EcalSimPulseShapeRcd& )
+{
+  std::cout<< "Produce EcalSimPulseShape set time interval="<<sim_pulse_shape_TI<< " ns"<< std::endl; 
+
+  auto result = std::make_unique<EcalSimPulseShape>();
+ 
+
+  result->time_interval = sim_pulse_shape_TI;
+
+  std::vector<double> EBshape_;
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: don't use a trailing `_` for local variables

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc [hunk 9]
```diff
@@ -3271,3 +3292,77 @@ EcalTrivialConditionRetriever::produceEcalSamplesCorrelation( const EcalSamplesC
...
+
+
+
+
+
+std::unique_ptr<EcalSimPulseShape>
+EcalTrivialConditionRetriever::getEcalSimPulseShapeFromConfiguration 
+( const EcalSimPulseShapeRcd& )
+{
+  std::cout<< "Produce EcalSimPulseShape set time interval="<<sim_pulse_shape_TI<< " ns"<< std::endl;
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: `cout` is not allowed, remove it everywhere (or use MessageLogger with appropriate severity level)

File: CondTools/Ecal/src/EcalDBCopy.cc [hunk 10]
```diff
@@ -456,6 +461,17 @@ else if (container == "EcalIntercalibConstantsMC") {
    std::cout << "sample mask pointer is: "<< obj<< std::endl;
    dbOutput->createNewIOV<const EcalSampleMask>( new EcalSampleMask(*obj),dbOutput->beginOfTime(), dbOutput->endOfTime(),recordName);
 
+ }  else if (container == "EcalSimPulseShape") {
+  std::cout << "sim pulse shape writing to DB... "<< std::endl;
```
Comment (kpedro88) [inline CondTools/Ecal/src/EcalDBCopy.cc]: remove cout or use message logger

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h (hunk 4 above)
Comment (kpedro88) [reply RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerWeights.h]: you're mixing tabs and spaces, try to avoid this

File: SimCalorimetry/EcalSimAlgos/interface/APDShape.h [hunk 11]
```diff
@@ -6,22 +6,11 @@
...
    public:
-  
-      APDShape( double tStart,
-		double tau     ) ;
-
-      ~APDShape() override ;
-
-      double threshold() const override ;
+   APDShape(bool useDB):EcalShapeBase(useDB){if(!useDB)buildMe();} // if useDB = true, then buildMe is executed when setEventSetup and DB conditions are available
+//   APDShape():EcalShapeBase(false){;}
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/interface/APDShape.h]: delete commented-out code

File: SimCalorimetry/EcalSimAlgos/interface/EBShape.h [hunk 12]
```diff
@@ -6,16 +6,11 @@
...
 {
    public:
-  
-      EBShape() ;
-
-      ~EBShape() override ;
-
-      double threshold() const override ;
+   EBShape(bool useDB):EcalShapeBase(useDB){if(!useDB)buildMe();} // if useDB = true, then buildMe is executed when setEventSetup and DB conditions are available 
+   //EBShape():EcalShapeBase(false){;}
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/interface/EBShape.h]: delete commented-out code

File: SimCalorimetry/EcalSimAlgos/interface/EEShape.h [hunk 13]
```diff
@@ -6,16 +6,11 @@
...
 {
    public:
-  
-      EEShape() ;
-
-      ~EEShape() override ;
-
-      double threshold() const override ;
+   EEShape(bool useDB):EcalShapeBase(useDB){if(!useDB)buildMe();} // if useDB = true, then buildMe is executed when setEventSetup and DB conditions are available}
+//   EEShape():EcalShapeBase(false){;}
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/interface/EEShape.h]: delete commented-out code

File: SimCalorimetry/EcalSimAlgos/interface/EcalShapeBase.h [hunk 14]
```diff
@@ -5,17 +5,29 @@
 //#include<stdexcept>
 #include "SimCalorimetry/CaloSimAlgos/interface/CaloVShape.h"
   
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h" 
+#include "CondFormats/EcalObjects/interface/EcalSimPulseShape.h"
+#include "CondFormats/DataRecord/interface/EcalSimPulseShapeRcd.h"
+
+#include <iostream>
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/interface/EcalShapeBase.h]: remove unneeded include

File: SimCalorimetry/EcalSimAlgos/src/EBShape.cc [hunk 15]
```diff
@@ -2,281 +2,278 @@
...
-//      shapeArray[i] = 10.82918*exp(-i/66.202 );
-   }
+	aVec = esps->barrel_shape; 
+	time_interval = esps->time_interval;
+	m_thresh = esps->barrel_thresh;
 
+    } else   // use old hardcoded arrays
+    {
+        time_interval = 1.0; 
+     	aVec.push_back(6.94068e-05);
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EBShape.cc]: fill this vector all at once: `aVec = {6.94068e-05, ...};`

File: SimCalorimetry/EcalSimAlgos/src/EBShape.cc [hunk 16]
```diff
@@ -2,281 +2,278 @@
...
-   aVec[241] = 0.286877 ; 
-   aVec[242] = 0.281831 ; 
-   aVec[243] = 0.276633 ; 
-   aVec[244] = 0.272283 ; 
-   aVec[245] = 0.268069 ; 
-   aVec[246] = 0.26399 ; 
-   aVec[247] = 0.258457 ; 
-   aVec[248] = 0.253549 ; 
-   aVec[249] = 0.249493 ; 
+	for( unsigned int i ( 250 ) ; i != 500 ; ++i ) aVec.push_back( exp(2.39735 - 0.0151053* ((double)i+1.0)) );
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EBShape.cc]: use a `reserve()` before filling in this loop

File: SimCalorimetry/EcalSimAlgos/src/EEShape.cc [hunk 17]
```diff
@@ -2,282 +2,276 @@
...
-   aVec[247] = 0.347136;
-   aVec[248] = 0.33872;
-   aVec[249] = 0.336706;
+	aVec = esps->endcap_shape; 
+	time_interval = esps->time_interval;
+	m_thresh = esps->endcap_thresh;
+    } else
+    {
+        time_interval = 1.0; 
+   	aVec.push_back(9.09091e-05);
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EEShape.cc]: fill vector all at once

File: SimCalorimetry/EcalSimAlgos/src/EEShape.cc [hunk 18]
```diff
@@ -2,282 +2,276 @@
...
+   	aVec.push_back(0.369457);
+   	aVec.push_back(0.365884);
+   	aVec.push_back(0.358889);
+   	aVec.push_back(0.353344);
+   	aVec.push_back(0.350364);
+   	aVec.push_back(0.347136);
+   	aVec.push_back(0.33872);
+   	aVec.push_back(0.336706);
 
+	for( unsigned int i ( 250 ) ; i != 500 ; ++i ) aVec.push_back( exp(2.63163e-01 - 1.37795e-02*(i - 151.0))  );
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EEShape.cc]: use reserve

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc [hunk 19]
```diff
@@ -2,27 +2,30 @@
...
 
 #include <cmath>
-#include <iostream>
-#include <fstream>
 #include <algorithm>
 
-const double EcalShapeBase::qNSecPerBin = 1./(1.*kNBinsPerNSec) ;
+#include <iostream>
+
+//const double EcalShapeBase::m_qNSecPerBin = 1./(1.*kNBinsPerNSec) ;
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: delete commented-out code

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc [hunk 20]
```diff
@@ -42,12 +45,27 @@ EcalShapeBase::timeToRise() const
 }
 
 
+double 
+EcalShapeBase::threshold()  const 
+{
+   return m_thresh              ;
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: remove unnecessary whitespace

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc [hunk 21]
```diff
@@ -42,12 +45,27 @@ EcalShapeBase::timeToRise() const
...
+   DVec shapeArray;
 
-   fillShape( shapeArray ) ;
+   float time_interval = 0;
+   fillShape(time_interval, m_thresh, shapeArray, m_es) ; // pure virtual function, implementation may vary for EB/EE/APD ...
+   m_arraySize = shapeArray.size();  // original data
+  
+   m_denseArraySize    = 10*m_arraySize; // dense array with interpolation between data 
+   m_kNBinsPerNSec     = (unsigned int) (10/time_interval); // used to be an unsigned int = 10 in  < CMSSW10X, should work for time intervals ~0.1, 0.2, 0.5, 1
+   m_qNSecPerBin = (1.*time_interval)/(10.);
```
Comment (kpedro88) [inline SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: why `1.*`?

File: SimCalorimetry/EcalSimAlgos/src/EEShape.cc (hunk 17 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EEShape.cc]: I am not sure I understand the requested vector fill, could you please provide an example in C++ ?

**Comment by davidlange6** [comment]
std::vector<int> v = {7, 5, 16, 8};

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc (hunk 6 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: done

File: CalibCalorimetry/EcalTrivialCondModules/interface/EcalTrivialConditionRetriever.h (hunk 7 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/interface/EcalTrivialConditionRetriever.h]: done

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc (hunk 8 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: OK

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc (hunk 9 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: done

File: CondTools/Ecal/src/EcalDBCopy.cc (hunk 10 above)
Comment (theofil) [reply CondTools/Ecal/src/EcalDBCopy.cc]: removed cout

File: SimCalorimetry/EcalSimAlgos/src/EEShape.cc (hunk 18 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EEShape.cc]: done

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc (hunk 19 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: done

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc (hunk 20 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: done

File: SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc (hunk 21 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EcalShapeBase.cc]: was not needed, removed the 1.*

File: SimCalorimetry/EcalSimAlgos/src/EEShape.cc (hunk 17 above)
Comment (theofil) [reply SimCalorimetry/EcalSimAlgos/src/EEShape.cc]: ok clarified by David, I believe the new codes uses the suggestion

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc [hunk 22]
```diff
@@ -3271,3 +3292,65 @@ EcalTrivialConditionRetriever::produceEcalSamplesCorrelation( const EcalSamplesC
        back_inserter(ipar->EEG1SamplesCorrelation));
   return ipar;
 }
+
+std::unique_ptr<EcalSimPulseShape>
+EcalTrivialConditionRetriever::getEcalSimPulseShapeFromConfiguration 
+( const EcalSimPulseShapeRcd& )
+{
+  std::cout<< "Produce EcalSimPulseShape set time interval="<<sim_pulse_shape_TI_<< " ns"<< std::endl;
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: remove cout

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc [hunk 23]
```diff
@@ -3271,3 +3292,65 @@ EcalTrivialConditionRetriever::produceEcalSamplesCorrelation( const EcalSamplesC
...
+  result->endcap_thresh = sim_pulse_shape_EE_thresh_;
+  result->apd_thresh = sim_pulse_shape_APD_thresh_;
+
+  copy(EBshape.begin(), EBshape.end(),
+       back_inserter(result->barrel_shape));
+  copy(EEshape.begin(), EEshape.end(),
+       back_inserter(result->endcap_shape));
+  copy(APDshape.begin(), APDshape.end(),
+       back_inserter(result->apd_shape));
+    std::cout<< "EcalSimPulseShape done ..."<< std::endl;
```
Comment (kpedro88) [inline CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: remove cout

**Review by kpedro88 — COMMENTED** [review]
In addition, there are a few comments still unaddressed:
https://github.com/cms-sw/cmssw/pull/23600#discussion_r199970114
https://github.com/cms-sw/cmssw/pull/23600#discussion_r199970130
https://github.com/cms-sw/cmssw/pull/23600#discussion_r199970186

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc (hunk 22 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: thanks for the catch!

File: CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc (hunk 23 above)
Comment (theofil) [reply CalibCalorimetry/EcalTrivialCondModules/src/EcalTrivialConditionRetriever.cc]: now fixed

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23600	9ca9de4
- changes since the last reco signoff were in packages not related to reco
```
CalibCalorimetry/EcalTrivialCondModules
CondFormats/EcalObjects
CondTools/Ecal
SimCalorimetry/EcalSimAlgos
```
- jenkins tests pass

**Comment by kpedro88** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by theofil** [comment]
could the assigned reviewers haven't yet signed this PR, please have a look ?

**Comment by rekovic** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
@jfernan2 @lpernie @arunhep could you please check it and comment or sign?

**Comment by lpernie** [comment]
+1

**Comment by fabiocos** [comment]
please test

**Comment by kpedro88** [comment]
+1
